File: src/main/java/com/rdvdev2/timetravelmod/impl/client/renderer/AnomalousAtemporalVoidBlockEntityRenderer.java
package com.rdvdev2.timetravelmod.impl.client.renderer;

import com.google.common.collect.ImmutableList;
import com.rdvdev2.timetravelmod.impl.common.block.entity.AnomalousAtemporalVoidBlockEntity;
import net.minecraft.block.Block;
import net.minecraft.client.render.RenderLayer;
import net.minecraft.client.render.VertexConsumer;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher;
import net.minecraft.client.render.block.entity.BlockEntityRenderer;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Matrix4f;

import java.util.List;
import java.util.Random;
import java.util.stream.IntStream;

public class AnomalousAtemporalVoidBlockEntityRenderer extends BlockEntityRenderer<AnomalousAtemporalVoidBlockEntity> {

    public AnomalousAtemporalVoidBlockEntityRenderer(BlockEntityRenderDispatcher dispatcher) {
        super(dispatcher);
    }

    private static final Random RANDOM = new Random(31100L);
    private static final List<RenderLayer> RENDER_LAYERS = IntStream.range(0, 16)
            .mapToObj(i -> RenderLayer.getEndPortal(i + 1))
            .collect(ImmutableList.toImmutableList());

    public void render(AnomalousAtemporalVoidBlockEntity entity, float f, MatrixStack matrixStack, VertexConsumerProvider vertexConsumerProvider, int i, int j) {
        RANDOM.setSeed(31100L);
        double distance = entity.getPos().getSquaredDistance(this.dispatcher.camera.getPos(), true);
        int iterations = this.calculateIterations(distance);
        Matrix4f matrix4f = matrixStack.peek().getModel();
        this.drawCube(entity, 0.15F, matrix4f, vertexConsumerProvider.getBuffer(RENDER_LAYERS.get(0)));

        for(int l = 1; l < iterations; ++l) {
            this.drawCube(entity, 2.0F / (float)(18 - l), matrix4f, vertexConsumerProvider.getBuffer(RENDER_LAYERS.get(l)));
        }

    }

    private void drawCube(AnomalousAtemporalVoidBlockEntity entity, float colorMultiplier, Matrix4f matrix4f, VertexConsumer vertexConsumer) {
        float r = (RANDOM.nextFloat() * 0.5F + 0.1F) * colorMultiplier;
        float g = (RANDOM.nextFloat() * 0.5F + 0.4F) * colorMultiplier;
        float b = (RANDOM.nextFloat() * 0.5F + 0.5F) * colorMultiplier;
        this.drawFace(entity, matrix4f, vertexConsumer, 0.0F, 1.0F, 0.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, r, g, b, Direction.SOUTH);
        this.drawFace(entity, matrix4f, vertexConsumer, 0.0F, 1.0F, 1.0F, 0.0F, 0.0F, 0.0F, 0.0F, 0.0F, r, g, b, Direction.NORTH);
        this.drawFace(entity, matrix4f, vertexConsumer, 1.0F, 1.0F, 1.0F, 0.0F, 0.0F, 1.0F, 1.0F, 0.0F, r, g, b, Direction.EAST);
        this.drawFace(entity, matrix4f, vertexConsumer, 0.0F, 0.0F, 0.0F, 1.0F, 0.0F, 1.0F, 1.0F, 0.0F, r, g, b, Direction.WEST);
        this.drawFace(entity, matrix4f, vertexConsumer, 0.0F, 1.0F, 0.0F, 0.0F, 0.0F, 0.0F, 1.0F, 1.0F, r, g, b, Direction.DOWN);
        this.drawFace(entity, matrix4f, vertexConsumer, 0.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, 0.0F, 0.0F, r, g, b, Direction.UP);
    }

    private void drawFace(AnomalousAtemporalVoidBlockEntity entity, Matrix4f matrix4f, VertexConsumer vertexConsumer, float x1, float x2, float y1, float y2, float z1, float z2, float z3, float z4, float r, float g, float b, Direction direction) {
        if (Block.shouldDrawSide(entity.getCachedState(), entity.getWorld(), entity.getPos(), direction)) {
            vertexConsumer.vertex(matrix4f, x1, y1, z1).color(r, g, b, 1.0F).next();
            vertexConsumer.vertex(matrix4f, x2, y1, z2).color(r, g, b, 1.0F).next();
            vertexConsumer.vertex(matrix4f, x2, y2, z3).color(r, g, b, 1.0F).next();
            vertexConsumer.vertex(matrix4f, x1, y2, z4).color(r, g, b, 1.0F).next();
        }
    }

    protected int calculateIterations(double d) {
        if (d > 36864.0D) {
            return 1;
        } else if (d > 25600.0D) {
            return 3;
        } else if (d > 16384.0D) {
            return 5;
        } else if (d > 9216.0D) {
            return 7;
        } else if (d > 4096.0D) {
            return 9;
        } else if (d > 1024.0D) {
            return 11;
        } else if (d > 576.0D) {
            return 13;
        } else {
            return d > 256.0D ? 14 : 15;
        }
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/client/renderer/AnomalousAtemporalVoidBlockEntityRenderer.kt
package com.rdvdev2.timetravelmod.impl.client.renderer

import com.rdvdev2.timetravelmod.impl.common.block.entity.AnomalousAtemporalVoidBlockEntity
import net.minecraft.block.Block
import net.minecraft.client.render.RenderLayer
import net.minecraft.client.render.VertexConsumer
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher
import net.minecraft.client.render.block.entity.BlockEntityRenderer
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.math.Direction
import net.minecraft.util.math.Matrix4f
import java.util.*
import java.util.stream.Collectors
import java.util.stream.IntStream

class AnomalousAtemporalVoidBlockEntityRenderer(dispatcher: BlockEntityRenderDispatcher?) : BlockEntityRenderer<AnomalousAtemporalVoidBlockEntity>(dispatcher) {

    override fun render(entity: AnomalousAtemporalVoidBlockEntity, f: Float, matrixStack: MatrixStack, vertexConsumerProvider: VertexConsumerProvider, i: Int, j: Int) {
        RANDOM.setSeed(31100L)
        val distance = entity.pos.getSquaredDistance(dispatcher.camera.pos, true)
        val iterations = calculateIterations(distance)
        val matrix4f = matrixStack.peek().model
        drawCube(entity, 0.15f, matrix4f, vertexConsumerProvider.getBuffer(RENDER_LAYERS[0]))
        for (l in 1 until iterations) {
            drawCube(entity, 2.0f / (18 - l).toFloat(), matrix4f, vertexConsumerProvider.getBuffer(RENDER_LAYERS[l]))
        }
    }

    private fun drawCube(entity: AnomalousAtemporalVoidBlockEntity, colorMultiplier: Float, matrix4f: Matrix4f, vertexConsumer: VertexConsumer) {
        val r = (RANDOM.nextFloat() * 0.5f + 0.1f) * colorMultiplier
        val g = (RANDOM.nextFloat() * 0.5f + 0.4f) * colorMultiplier
        val b = (RANDOM.nextFloat() * 0.5f + 0.5f) * colorMultiplier
        drawFace(entity, matrix4f, vertexConsumer, 0.0f, 1.0f, 0.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, r, g, b, Direction.SOUTH)
        drawFace(entity, matrix4f, vertexConsumer, 0.0f, 1.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, r, g, b, Direction.NORTH)
        drawFace(entity, matrix4f, vertexConsumer, 1.0f, 1.0f, 1.0f, 0.0f, 0.0f, 1.0f, 1.0f, 0.0f, r, g, b, Direction.EAST)
        drawFace(entity, matrix4f, vertexConsumer, 0.0f, 0.0f, 0.0f, 1.0f, 0.0f, 1.0f, 1.0f, 0.0f, r, g, b, Direction.WEST)
        drawFace(entity, matrix4f, vertexConsumer, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 1.0f, 1.0f, r, g, b, Direction.DOWN)
        drawFace(entity, matrix4f, vertexConsumer, 0.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 0.0f, 0.0f, r, g, b, Direction.UP)
    }

    private fun drawFace(entity: AnomalousAtemporalVoidBlockEntity, matrix4f: Matrix4f, vertexConsumer: VertexConsumer, x1: Float, x2: Float, y1: Float, y2: Float, z1: Float, z2: Float, z3: Float, z4: Float, r: Float, g: Float, b: Float, direction: Direction) {
        if (Block.shouldDrawSide(entity.cachedState, entity.world, entity.pos, direction)) {
            vertexConsumer.vertex(matrix4f, x1, y1, z1).color(r, g, b, 1.0f).next()
            vertexConsumer.vertex(matrix4f, x2, y1, z2).color(r, g, b, 1.0f).next()
            vertexConsumer.vertex(matrix4f, x2, y2, z3).color(r, g, b, 1.0f).next()
            vertexConsumer.vertex(matrix4f, x1, y2, z4).color(r, g, b, 1.0f).next()
        }
    }

    private fun calculateIterations(d: Double): Int {
        return if (d > 36864.0) 1
        else if (d > 25600.0) 3
        else if (d > 16384.0) 5
        else if (d > 9216.0) 7
        else if (d > 4096.0) 9
        else if (d > 1024.0) 11
        else if (d > 576.0) 13
        else if (d > 256.0) 14
        else 15
    }

    companion object {
        private val RANDOM = Random(31100L)
        private val RENDER_LAYERS: List<RenderLayer> = IntStream.range(0, 16)
            .mapToObj { i: Int -> RenderLayer.getEndPortal(i + 1) }
            .collect(Collectors.toUnmodifiableList())
    }
}